### PR TITLE
Update TF text classification example

### PR DESCRIPTION
**Pull request by Rocketknight1**
This updates the TF text classification example with several refactors, as well as multi-GPU and TPU support.

It's late so I'd like to do one more pass over everything before merging tomorrow, but I'm opening for reviews before I head off for the evening!

### CODE REVIEW

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 1]
```diff
@@ -119,6 +60,56 @@ def on_epoch_end(self, epoch, logs=None):
         self.model.save_pretrained(self.output_dir)
 
 
+def convert_dataset_for_tensorflow(dataset, non_label_column_names, batch_size, labels, dataset_mode, drop_remainder):
```
Comment (sgugger) [inline examples/tensorflow/text-classification/run_text_classification.py:63]: dataset_mode and drop_remainder could have defaults here.

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 2]
```diff
@@ -119,6 +60,56 @@ def on_epoch_end(self, epoch, logs=None):
...
+    """Converts a Hugging Face dataset to a Tensorflow Dataset or Sequence object. We usually only want a Dataset when
+    we're training on TPU, as we get the nice feature of variable-length batches when we put the data in a Sequence
+    object instead."""
```
Comment (sgugger) [inline examples/tensorflow/text-classification/run_text_classification.py:66]: For the docstrings on multiple line we use the following:
```suggestion
    """
    Converts a Hugging Face dataset to a Tensorflow Dataset or Sequence object. We usually only want a Dataset when
    we're training on TPU, as we get the nice feature of variable-length batches when we put the data in a Sequence
    object instead.
    """
```

Comment (sgugger) [reply examples/tensorflow/text-classification/run_text_classification.py:66]: It also looks like we never return a `Sequence`?

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 3]
```diff
@@ -398,10 +358,12 @@ def main():
             sentence1_key, sentence2_key = non_label_column_names[:2]
         else:
             sentence1_key, sentence2_key = non_label_column_names[0], None
-
     # Padding strategy
     if data_args.pad_to_max_length:
         padding = "max_length"
+    # elif isinstance(training_args.strategy, tf.distribute.TPUStrategy):
+    #    logger.warning("Forcing --pad_to_max_length because we're running on TPU!")
+    #     padding = "max_length"
```
Comment (sgugger) [inline examples/tensorflow/text-classification/run_text_classification.py:366]: Why commented out?

File: examples/tensorflow/text-classification/run_text_classification.py [hunk 4]
```diff
@@ -452,85 +414,129 @@ def preprocess_function(examples):
...
+            "test": data_args.max_test_samples,
+        }
+        for key in ("train", "validation", "test"):
+            if key not in datasets:
+                tf_data[key] = None
+                continue
+            if key in ("train", "validation"):
+                use_labels = True
+            else:
+                use_labels = False
```
Comment (sgugger) [inline examples/tensorflow/text-classification/run_text_classification.py:478]: We could have a test set with labels (people want to evaluate on the test set of GLUE for instance) so the test should be if there is a `"labels"` columns IMO.

**Review by sgugger — APPROVED** [review]
Thanks for tackling this! I left a few comments but it's looking good!

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 2 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: You're right - this text has been updated

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 3 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py:353]: This code is a leftover from when we padded the entire dataset into one dense Tensor. This is the least memory-efficient option of all, and we don't need to do it even on TPU. Instead, we can just ensure that we pad batches to the same size each time, which is now handled in the Dataset logic. I've removed it!

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 4 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py:465]: This is a good point - right now the code is really treating the 'test' set as the set the user wants predictions on, and uses 'validation' as the set to compute accuracy and metrics on. This is probably not very clear. Would it make more sense to have a separate --prediction_inputs file or something like that?

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 1 above)
Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: Done!

File: examples/tensorflow/text-classification/run_text_classification.py (hunk 4 above)
Comment (sgugger) [reply examples/tensorflow/text-classification/run_text_classification.py:461]: I would align to the PyTorch script that returns:
- the predictions if no labels
- the predictions + metrics if labels

Comment (Rocketknight1) [reply examples/tensorflow/text-classification/run_text_classification.py]: Done!

File: examples/tensorflow/text-classification/README.md [hunk 5]
```diff
@@ -54,6 +54,20 @@ After training, the model will be saved to `--output_dir`. Once your model is tr
...
+### Multi-GPU and TPU usage
+
+By default, the script uses a `MirroredStrategy` and will use multiple GPUs effectively if they are available. TPUs
+can also be used by passing the name of the TPU resource with the `--tpu` argument.
+
+### Memory usage and data loading
+
+One thing to note is that all data is loaded into memory in this script. Most text classification datasets are small
+enough that this is not an issue, but if you have a very large dataset you will need to modify the script to handle
+data streaming,. This is particularly challenging for TPUs, given the stricter requirements and the sheer volume of data
```
Comment (LysandreJik) [inline examples/tensorflow/text-classification/README.md:66]:
```suggestion
data streaming. This is particularly challenging for TPUs, given the stricter requirements and the sheer volume of data
```

**Review by LysandreJik — APPROVED** [review]
Looks great, thanks for working on this @Rocketknight1!
